R/decisiontreeutils.R

```diff
@@ -0,0 +1,363 @@
+#' Error checking pre-model
+#' Does not return anything - just throws errror
+#'
+#' @param config list of config options
+#' @param the.data incoming data
+preModelCheckDT <- function(config, the.data) {
+  data_names <- names(the.data)
+  names_list <- getNamesFromOrdered(config$used.weights, data_names)
+  name_y_var <- names_list$y
+  cp <- ifelse(config$the.cp == "Auto" || config$the.cp == "", .00001, config$the.cp)
+
+
+  the_target <- the.data[[name_y_var]]
+  if (is.numeric(the_target) && length(unique(the_target)) < 5 && !is_XDF) {
+    AlteryxMessage("The target variable is numeric, however, it has 4 or fewer unique values.", iType = 2, iPriority = 3)
+  }
+
+  if(rpart_params$cp < 0 || rpart_params$cp > 1) {
+    stop.Alteryx("The complexity parameter must be between 0 and 1. Please try again.")
+  }
+
+  if(is.na(as.numeric(config$the.cp)) && !(config$the.cp == "Auto" || config$the.cp == "")) {
+    stop.Alteryx("The complexity parameter provided is not a number. Please enter a new value and try again.")
+  }
+}
+
+#' Creation of components for model object evaluation
+#'
+#' @param config list of config options
+#' @param data list of datastream inputs
+#' @return list with components needed to create model
+createDTParams <- function(config, data) {
+  # use lists to hold params for rpart and rxDTree functions
+  rpart_params <- rxDTree_params <- list()
+
+  # Determine if a compute context is being used, and whether it is XDF
+  # The mechanism for doing this depends on how the information is transfered
+  xdf_properties <- getXdfProperties("#1")
+  is_XDF <- xdf_properties$is_XDF
+  xdf_path <- xdf_properties$xdf_path
+
+  # get data param
+  the.data <- data$data_stream1
+  data_names <- names(the.data)
+  rpart_params$data <- quote(the.data)
+  rxDTree_params$data <- quote(xdf_path)
+
+  # Get the field names
+  names_list <- getNamesFromOrdered(config$used.weights, data_names)
+  name_y_var <- names_list$y
+  names_x_vars <- names_list$x
+  name_weight_var <- names_list$w
+
+  # use field names to get formula param
+  dtree_formula <- makeFormula(names_x_vars, name_y_var)
+  rpart_params$formula <- rxDTree_params$formula <- dtree_formula
+
+  # get weights param
+  weight_arg <- ifelse(config$used.weights, name_weight_var, NULL)
+  rpart_params$weights <- rxDTree_params$pweights <- weight_arg
+
+  # get method and parms params
+  if(config$select.type) {
+    if(config$classification) {
+      rpart_params$method <- rxDTree_params$method <- "class"
+      if(config$use.gini) {
+        rpart_params$parms <- rxDTree_params$parms <- quote(list(split = "gini"))
+      } else {
+        rpart_params$parms <- rxDTree_params$parms <- quote(list(split = "information"))
+      }
+    } else {
+      rpart_params$method <- rxDTree_params$method <- "anova"
+    }
+  }
+
```

Comment:
It would be good to abstract these out into functions. The advantage would be that you could write unit tests that ensure correctness. It also makes the code more readable, especially if you give the function an appropriate name.

```r
#' Add method and parms to the parameter list
#' 
#' @param config configuration provided to the tool
#' @param params parameter list
addMethodandParms <- function(params, config){
  if(config$select.type) {
    if(config$classification) {
      args$method <- "class"
      if(config$use.gini) {
        params$parms <- quote(list(split = "gini"))
      } else {
        params$parms <- quote(list(split = "information"))
      }
    } else {
      params$method <- rxDTree_params$method <- "anova"
    }
  }
  return (params)
}  
```

Also note the ordering of arguments. By putting the argument that gets modified first, you can take advantage of using pipes.

```
params %>%
   addMethodAndParms %>%
   ...
```